engine/evarutil.ml

```diff
@@ -721,6 +726,33 @@ let rec advance sigma evk =
       | Some evk -> advance sigma evk
       | None -> None
 
+let reachable_from_evars current_sigma evars evk =
+  let rec search evk' visited =
+    if Evar.Set.mem evk' visited then visited
+    else
+      let visited = Evar.Set.add evk' visited in
+      let evi = Evd.find current_sigma evk' in
+      match Evd.evar_body evi with
+      | Evd.Evar_empty -> visited
+      | Evd.Evar_defined c ->
+        match is_restricted_evar current_sigma evk' with
+        | Some evk' ->
+          if Evar.equal evk' evk then raise Exit
+          else search evk' visited
+        | None ->
+          if is_evar_evar_solution current_sigma evk' then begin
```

Comment:
@mattam82 I didn't understand well why this metadata is necessary. Why can't we just look at the instantiated body instead and see if it is an evar?

Reply:
I think we want to distinguish evar-evar solutions produced by unification (make it transparent to the user), while not messing with users explicit  instantiations. But you might be right that this is actually overkill! There must be a trickier reason...

Reply:
I mean keeping the metadata vs checking if the body is an evar might have the same semantics in general, if any instantiation goes through solve_evar_evar

Reply:
evar-evar problems can also be solved by `solve_simple_eqn` AFAIK, but I don't know if it is a bug to distinguish it or if that's what you wanted to do.

Reply:
Although maybe it still goes through `solve_evar_evar` in the end, I'd have to check.